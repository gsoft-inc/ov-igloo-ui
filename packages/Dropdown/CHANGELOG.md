# @igloo-ui/dropdown

## 1.4.1

### Patch Changes

- 89d10c5: Updated recharts version and fixed dropdown mock

## 1.4.0

### Minor Changes

- 9d9d72a: Allow Combobox, Select, TagPicker and Dropdown components to scroll if it's too long and runs out of room in the viewport. New properties added to Dropdown are: isReferenceWidth, isScrollable and footer. The display of the footer was moved to the dropdown instead of leaving it in combobox because the dropdown handles the scroll styles and the footer is sticky.

## 1.3.0

### Minor Changes

- 6e2e96a: Added the disablePortal prop for when you don't want the dropdown to be appended to the end of the body. False by default so it won't break the current behavior.
- 6e2e96a: The dropdown will now follow the reference element if it changes positions without screen resize or scoll.

## 1.2.0

### Minor Changes

- 0b8a62f: The dropdown will now follow the reference element if it changes positions without screen resize or scoll.

## 1.1.4

### Patch Changes

- a41e1d4: Fix: moved destructuring of props to function parameters so that storybook will automatically show the default value in the props table. It was done in Modal because it needed to send props to useDialog. In this case, @default annotation was used in comments.

## 1.1.3

### Patch Changes

- bf478b5: Added missing tests. Fixed missing code in stories when clicking 'Show code'. Updated Dropdown mock. Added new optional prop highlightToday to help with test snapshots.

## 1.1.2

### Patch Changes

- 3f50827: Made dropdown portal to the end of the document so that it works in Modals. Removed unused line in Modal.

## 1.1.1

### Patch Changes

- 7dcba93: Initial release of List and Combobox components. Select was updated to use the Dropdown component that was created previously as well as the new List component.

## 1.1.0

### Minor Changes

- 6b869be: Updated jsPopper to floating-ui and Fixed the weird positiong behavior of the dropdown

## 1.0.0

### Major Changes

- 9b9ec61: Updated the way the dropdown list displays. It is appended to the body.

  ## BREAKING CHANGE

  The dropdown uses the usePopper hook from react popper to allow the dropdown to be appended to the body. This prevents many potential bugs such as creating a scrollbar in a div when the dropdown should display outside the div.
  Default position looks the same, but was updated to bottomLeft. Position 'bottom' used to look like 'bottomLeft'. It was pointless to have two positions do the same thing so now 'bottom' is horizontally positioned in the center to be aligned with js popper positioning.

## 0.2.1

### Patch Changes

- 6ee9c45: Updated anything causing errors or warnings in the browser console

## 0.2.0

### Minor Changes

- 64ca515: add new position: `topLeft`, `topRight`, `bottomLeft`, `bottomRight`

## 0.1.1

### Patch Changes

- 0640409: Update packages

  - "react-aria": "^3.21.0",
  - "@react-aria/link": "^3.35",
  - "react-stately": "^3.19.0"

## 0.1.0

### Minor Changes

- 9842a0d: Initial release of the Dropdown component
