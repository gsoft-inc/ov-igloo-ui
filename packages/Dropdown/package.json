{
    "name": "@igloo-ui/dropdown",
    "version": "1.8.0",
    "main": "dist/Dropdown.js",
    "module": "dist/Dropdown.js",
    "types": "dist/Dropdown.d.ts",
    "styles": "dist/dropdown.css",
    "publishConfig": {
        "access": "public"
    },
    "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
    "author": "Officevibe Inc.",
    "contributors": [
        "Franck franck.gaudin@gsoft.com"
    ],
    "license": "Apache-2.0",
    "peerDependencies": {
        "react": ">= 16.8.6"
    },
    "files": [
        "dist"
    ],
    "scripts": {
        "build": "rollup -c rollup.config.js"
    },
    "dependencies": {
        "@floating-ui/react": "^0.24.8",
        "@igloo-ui/tokens": "^2.0.0",
        "@hopper-ui/tokens": "^3.1.4",
        "classnames": "^2.3.2"
    },
    "browserslist": [
        "> 1%",
        "last 2 versions",
        "IE >= 11"
    ]
}
