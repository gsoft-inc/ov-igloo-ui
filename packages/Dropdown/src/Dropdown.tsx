import * as React from "react";
import cx from "classnames";
import {
    flip,
    size as fuiSize,
    useMergeRefs,
    offset,
    hide,
    autoUpdate,
    useFloating,
    FloatingFocusManager,
    useDismiss,
    useInteractions,
    useTransitionStyles,
    useRole,
    FloatingPortal
} from "@floating-ui/react";

import "./dropdown.scss";

export type Position =
  | "top"
  | "top-start"
  | "top-end"
  | "bottom"
  | "bottom-start"
  | "bottom-end";
export type Size = "xsmall" | "small" | "medium" | "large";
export type Role = "listbox" | "menu";

export interface DropdownProps
    extends Omit<React.ComponentPropsWithRef<"div">, "content"> {
    /** The target button, text, svg etc.. of the Dropdown */
    children?: React.ReactElement;
    /** Add a specific class to the dropdown */
    className?: string;
    /** Default value displayed in the Dropdown. */
    content: React.ReactNode;
    /** Disables the component from appending the dropdown
   * to the end of the body using ReactPortal */
    disablePortal?: boolean;
    /** Position of the Dropdown */
    position?: Position;
    /** Changes the size of the card, giving it more or less padding */
    size?: Size;
    /** True if the Dropdown list is displayed */
    isOpen?: boolean;
    /** Callback when the user clicks outside the Dropdown. */
    onClose?: () => void;
    /** Callback when the Dropdown is closed and animations are done */
    onAfterClose?: () => void;
    /** Add a data-test tag for automated tests */
    dataTest?: string;
    /** The role of the dropdown for aria purposes */
    role?: Role;
    /** Render the reference element to be able to add the
   * reference props directly. This overrides children */
    renderReference?: (
        props: React.HTMLProps<HTMLButtonElement>,
    ) => React.ReactElement;
    /** Whether or not the dropdown should take the
   * width of the reference element */
    isReferenceWidth?: boolean;
    /** Whether or not the dropdown should be scrollable */
    isScrollable?: boolean;
    /** The footer content of the dropdown */
    footer?: React.ReactNode;
    /** Callback when the user scrolls to the end of the Dropdown */
    onScrollEnd?: () => void;
    /** The threshold in pixels from the bottom of the dropdown
   * to trigger the onScrollEnd callback */
    scrollEndThreshold?: number;
}

const Dropdown: React.FunctionComponent<DropdownProps> = React.forwardRef(
    (
        {
            children,
            className,
            content,
            size = "xsmall",
            onClose,
            onAfterClose,
            dataTest,
            disablePortal = false,
            isOpen = false,
            position = "bottom-start",
            style,
            role = "listbox",
            renderReference,
            isReferenceWidth = false,
            isScrollable = false,
            footer,
            onScrollEnd,
            scrollEndThreshold = 30,
            ...rest
        }: DropdownProps,
        ref: React.ForwardedRef<HTMLDivElement>
    ) => {
        const [dropdownPreviouslyOpen, setDropdownPreviouslyOpen] =
      React.useState(false);

        const [scrollEndTriggered, setScrollEndTriggered] = React.useState(false);

        const handleOnClose = (): void => {
            if (onClose) {
                onClose();
            }
        };

        const handleOpenChange = (open: boolean): void => {
            if (!open) {
                handleOnClose();
            }
        };

        const onDropdownScroll = (e: React.UIEvent<HTMLDivElement>): void => {
            const { scrollHeight, scrollTop, clientHeight } = e.currentTarget;
            const threshold = scrollEndThreshold || 1;

            if (scrollEndTriggered) {
                if (scrollHeight - scrollTop - clientHeight > threshold) {
                    setScrollEndTriggered(false);
                }

                return;
            }

            if (scrollHeight - scrollTop - clientHeight <= threshold) {
                if (onScrollEnd) {
                    setScrollEndTriggered(true);
                    onScrollEnd();
                }
            }
        };

        const { x, y, strategy, refs, context, middlewareData } = useFloating({
            placement: position,
            open: isOpen,
            strategy: disablePortal ? "absolute" : "fixed",
            onOpenChange: handleOpenChange,
            whileElementsMounted: (referenceEl, floatingEl, updatePosition) => {
                const cleanup = autoUpdate(referenceEl, floatingEl, updatePosition, {
                    animationFrame: true
                });

                return cleanup;
            },
            middleware: [
                offset(1),
                flip({ padding: 10 }),
                fuiSize({
                    apply({ rects, elements, availableHeight }) {
                        const styleProps: React.CSSProperties = {};
                        if (isScrollable) {
                            styleProps.maxHeight = `${Math.max(60, availableHeight)}px`;
                        }
                        if (isReferenceWidth) {
                            styleProps.width = `${rects.reference.width}px`;
                        }
                        Object.assign(elements.floating.style, styleProps);
                    },
                    padding: 10
                }),
                hide({
                    padding: 10
                })
            ]
        });

        const mergedDropdownRef = useMergeRefs([refs.setFloating, ref]);

        const { getReferenceProps, getFloatingProps } = useInteractions([
            useDismiss(context),
            useRole(context, { role })
        ]);

        const { isMounted, styles } = useTransitionStyles(context, {
            duration: 150,
            initial: ({ side }) => ({
                opacity: 0,
                transform: side === "bottom" ? "translateY(1rem)" : "translateY(-1rem)"
            }),
            open: {
                opacity: 1,
                transform: "translateY(0rem)"
            },
            close: ({ side }) => ({
                opacity: 0,
                transform: side === "bottom" ? "translateY(1rem)" : "translateY(-1rem)"
            })
        });

        const dropdownClasses = cx("ids-dropdown", className, {
            [`ids-dropdown--${size}`]: size !== "xsmall",
            [`ids-dropdown--${position}`]: position !== "bottom",
            "ids-dropdown--hidden": middlewareData.hide?.referenceHidden,
            "ids-dropdown--scrollable": isScrollable
        });
        const dropdownScrollProps: React.HTMLAttributes<HTMLDivElement> = {
            className: "ids-dropdown__scroll-content"
        };
        if (onScrollEnd) {
            dropdownScrollProps.onScroll = onDropdownScroll;
        }

        const floatingElem = isMounted && (
            <FloatingFocusManager context={context} modal={false} initialFocus={-1}>
                <div
                    ref={mergedDropdownRef}
                    className={dropdownClasses}
                    data-test={dataTest}
                    {...rest}
                    data-show={isOpen}
                    style={{
                        position: strategy,
                        top: y ?? 0,
                        left: x ?? 0,
                        ...styles,
                        ...style
                    }}
                    {...getFloatingProps()}
                >
                    <div {...dropdownScrollProps}>{content}</div>
                    {footer && <div className="ids-dropdown__footer">{footer}</div>}
                </div>
            </FloatingFocusManager>
        );

        React.useEffect(() => {
            if (isMounted && !dropdownPreviouslyOpen) {
                setDropdownPreviouslyOpen(true);
            }
            if (!isMounted && dropdownPreviouslyOpen && onAfterClose) {
                onAfterClose();
            }
        }, [isMounted, onAfterClose, dropdownPreviouslyOpen]);

        return (
            <>
                {renderReference ? (
                    renderReference({ ref: refs.setReference, ...getReferenceProps() })
                ) : (
                    <div
                        className="ids-dropdown__ref"
                        ref={refs.setReference}
                        {...getReferenceProps()}
                    >
                        {children}
                    </div>
                )}
                {disablePortal ? (
                    floatingElem
                ) : (
                    <FloatingPortal>{floatingElem}</FloatingPortal>
                )}
            </>
        );
    }
);

export default Dropdown;
