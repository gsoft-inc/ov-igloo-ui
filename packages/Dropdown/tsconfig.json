{
  "extends": "../../tsconfig.build.json",
  "include": [
    "src/**/*",
    "../../typings/**/*",
    "rollup.config.js",
    "__mocks__"
  ],
  "exclude": ["dist"],
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "declarationDir": "./dist",
    "composite": true,
    "baseUrl": ".",
    "typeRoots": ["../../typings", "../../node_modules/@types"]
  }
}
