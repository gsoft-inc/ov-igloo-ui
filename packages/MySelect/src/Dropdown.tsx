import * as React from 'react';
import cx from 'classnames';

import './dropdown.scss';

export enum DropdownPositions {
  Top,
  Right,
  Bottom,
  Left,
}

export interface DropdownProps extends React.ComponentPropsWithRef<'div'> {
  /** Default value displayed in the dropdown. */
  children?: React.ReactNode;
  /** Position of the dropdown */
  position?: DropdownPositions;
}

const Dropdown: React.FunctionComponent<DropdownProps> = React.forwardRef(
  (props: DropdownProps, ref: React.Ref<HTMLDivElement>) => {
    const { children, position = DropdownPositions.Bottom, ...rest } = props;

    const dropdownClasses = cx('ids-dropdown', {
      'ids-dropdown--top': position === DropdownPositions.Top,
      'ids-dropdown--right': position === DropdownPositions.Right,
      'ids-dropdown--bottom': position === DropdownPositions.Bottom,
      'ids-dropdown--left': position === DropdownPositions.Left,
    });

    return (
      <div ref={ref} className={dropdownClasses} {...rest}>
        {children}
      </div>
    );
  }
);

export default Dropdown;
