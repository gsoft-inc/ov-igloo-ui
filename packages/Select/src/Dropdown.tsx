import * as React from 'react';
import cx from 'classnames';
import { useTransition, animated } from 'react-spring';

import './dropdown.scss';

export enum DropdownPositions {
  Top,
  Right,
  Bottom,
  Left,
}

const DeterminePosition = (position: DropdownPositions): string => {
  if (position === DropdownPositions.Right) {
    return 'translateX(-1rem)';
  }

  if (position === DropdownPositions.Top) {
    return 'translateY(1rem)';
  }

  return 'translateY(-1rem)';
};

export interface DropdownProps extends React.ComponentPropsWithRef<'div'> {
  /** Default value displayed in the Dropdown. */
  children?: React.ReactNode;
  /** Position of the Dropdown */
  position?: DropdownPositions;
  /** True if the Dropdown list is displayed. */
  isOpen?: boolean;
}

const Dropdown: React.FunctionComponent<DropdownProps> = React.forwardRef(
  (props: DropdownProps, ref: React.Ref<HTMLDivElement>) => {
    const {
      children,
      position = DropdownPositions.Bottom,
      isOpen = false,
      ...rest
    } = props;

    const initialTransform = DeterminePosition(position);
    const isOnTheRight = position === DropdownPositions.Right;
    const transitions = useTransition(isOpen, {
      from: { opacity: 0, transform: initialTransform },
      enter: {
        opacity: 1,
        transform: isOnTheRight ? 'translateX(0rem)' : 'translateY(0rem)',
      },
      leave: { opacity: 0, transform: initialTransform },
      config: { duration: 150 },
    });

    const dropdownClasses = cx('ids-dropdown', {
      'ids-dropdown--top': position === DropdownPositions.Top,
      'ids-dropdown--right': position === DropdownPositions.Right,
      'ids-dropdown--bottom': position === DropdownPositions.Bottom,
      'ids-dropdown--left': position === DropdownPositions.Left,
    });

    return transitions((styles) => (
      <animated.div
        style={styles}
        ref={ref}
        className={dropdownClasses}
        {...rest}
      >
        {children}
      </animated.div>
    ));
  }
);

export default Dropdown;
