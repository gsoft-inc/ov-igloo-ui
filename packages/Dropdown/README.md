# Dropdown

Displays a sub-element overlaid and renders the content

<Example />

<ReferenceLinks />

## Installation

To install `@igloo-ui/dropdown` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/dropdown
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/dropdown
```

## Usage

Then to use the component in your code just import it!

```jsx
import Dropdown from '@igloo-ui/dropdown';
import Button from '@igloo-ui/button';
import Datepicker from '@igloo-ui/datepicker';

const [show, setShow] = React.useState(false);

<Dropdown
  isOpen={show}
  onClose={() => setShow(false)}
  content={<Datepicker selectedDay="2022-12-12" />}
>
  <Button appearance="secondary" size="small" onClick={() => setShow(!show)}>
    Chose date
  </Button>
</Dropdown>;
```
