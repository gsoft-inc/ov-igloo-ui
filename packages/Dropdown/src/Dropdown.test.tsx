/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen } from '@testing-library/react';

import Dropdown from './Dropdown';

describe('Dropdown', () => {
  test('It should render without error and a snapshot', () => {
    const { asFragment } = render(
      <Dropdown content="All settings" dataTest="dropdown1">
        <button>Settings</button>
      </Dropdown>
    );

    expect(screen.getByTestId('dropdown1')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
