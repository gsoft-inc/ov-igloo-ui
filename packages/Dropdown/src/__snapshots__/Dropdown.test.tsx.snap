// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown It should render without error and a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-dropdown__ref"
  >
    <button>
      Settings
    </button>
  </div>
</DocumentFragment>
`;
